### ISSUE TITLE
Masking out Loads from mask_modded out regions

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#150321]
# Summary
There is a very common gotcha/footgun that users run into when composing score_mods w/ blockmasks. 

```Py
import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention

B, H, SEQ_LEN, HEAD_DIM = 1, 1, 128, 16
MAX_LEN = 127 # 1 less then possible values

buffer = torch.arange(MAX_LEN).to(device="cuda")
def make_tensor():
    return torch.randn((B, SEQ_LEN, H, HEAD_DIM), device="cuda").permute(0,2,1,3)
q, k, v = make_tensor(), make_tensor(), make_tensor()

def causal(b, h, q, kv):
    return q >= kv

def mask_mod(b, h, q, kv):
    upper = kv < MAX_LEN
    return upper & causal(b, h, q, kv)

def score_mod(score, b, h, q, kv):
    #  Users expect that we will only apply where we have masked out values
    # However for the last block and last row we will attempt to read from the buffer
    return score + buffer[kv]

bm = create_block_mask(mask_mod, None, None, SEQ_LEN, SEQ_LEN)
out = flex_attention(q, k, v, score_mod=score_mod, block_mask=bm)
print(out)
```

Above is a minimal repro that highlights this issue. The crux of the issues has to do w/ the overlapping semantics of mask_mods and score_mods. If a users is reading from a buffer and utilizing both a score_mod and mask_mod they typically ensure that any read is globally correct by using the mask_mod to "mask out" any invalid locations.  

In this example the line doing that is: `upper = kv < MAX_LEN`. And now they write their score_mod in a mask oblivious way - which feels natural.
